### PR TITLE
Fix memory leakage when port-forwarding

### PR DESCRIPTION
**Pull request by bitoku**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

This PR forces cri-o to drain the stream when it finishes port-forward in order to avoid failure to close the connection which causes memory leakage.

The issue details are here.
https://issues.redhat.com/browse/OCPBUGS-30978
> while sending multiple failing port-forwarding requests, crio memory usage increases until the node dies

It had been reported before (https://bugzilla.redhat.com/show_bug.cgi?id=1798193) and fixed in https://github.com/cri-o/cri-o/commit/7016c3e13fe19272d6a680ca2172df02da75a35f.
However at some point, it regressed.

#### Which issue(s) this PR fixes:

Fixes https://issues.redhat.com/browse/OCPBUGS-30978

#### Does this PR introduce a user-facing change?

```release-note
Fix memory leakage when sending a failing port-forward request
```

### CODE REVIEW

File: server/container_portforward.go [hunk 1]
```diff
@@ -25,17 +25,19 @@ func (s StreamService) PortForward(ctx context.Context, podSandboxID string, por
...
-				log.Errorf(ctx, "Error closing port forward stream after other error: %v", copyError)
-			}()
+		if stream == nil {
+			return
 		}
+		go func() {
+			_, err := pools.Copy(io.Discard, stream)
+			if err != nil {
+				log.Errorf(ctx, "could not drain the stream with the error: %v", err)
+			}
```
Comment (kwilczynski) [inline server/container_portforward.go:39]:
```suggestion
			if _, err := pools.Copy(io.Discard, stream); err != nil {
				log.Errorf(ctx, "Unable to drain the stream data: %v", err)
			}
```

File: server/container_portforward.go [hunk 2]
```diff
@@ -25,17 +25,19 @@ func (s StreamService) PortForward(ctx context.Context, podSandboxID string, por
...
+	// ref https://issues.redhat.com/browse/OCPBUGS-30978
 	defer func() {
-		if emptyStreamOnError && stream != nil {
-			go func() {
-				_, copyError := pools.Copy(io.Discard, stream)
-				log.Errorf(ctx, "Error closing port forward stream after other error: %v", copyError)
-			}()
+		if stream == nil {
+			return
 		}
```
Comment (kwilczynski) [inline server/container_portforward.go:34]: If we keep the if statement, but remote the `emptyStreamOnError` from it, then we have the smallest changeset.

Comment (haircommander) [reply server/container_portforward.go:34]: I like the updated structure

File: server/container_portforward.go [hunk 3]
```diff
@@ -25,17 +25,19 @@ func (s StreamService) PortForward(ctx context.Context, podSandboxID string, por
...
+		go func() {
+			_, err := pools.Copy(io.Discard, stream)
+			if err != nil {
```
Comment (haircommander) [inline server/container_portforward.go:37]: nit: this can be compressed to one line

File: server/container_portforward.go (hunk 1 above)
Comment (bitoku) [reply server/container_portforward.go]: thanks!

File: server/container_portforward.go (hunk 2 above)
Comment (bitoku) [reply server/container_portforward.go:34]: If there is no objections about it, I'll keep the update one. thanks!

**Comment by haircommander** [comment]
/approve

thank you!

**Pull request merged**